Abstracts out ShopBasedCartContext::reset to an interface

palttamas opened the pull request:
`ShopBasedCartContext` now implements the new `ResettingCartContextInterface`.
    
The service `sylius.context.cart.new_shop_based` is no longer explicitly tagged with the `kernel.reset` tag.
    
Cart contexts that implement the `ResettingCartContextInterface` will be automatically tagged with the `kernel.reset` tag.


| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12 <!-- see the comment below -->                  |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                      |
| BC breaks?      | no                                                       |
| Deprecations?   | no <!-- don't forget to update the UPGRADE-*.md file --> |
| Related tickets | [fixes #15021 ](https://github.com/Sylius/Sylius/issues/15021)                    |
| License         | MIT                                                          |

<!--
 - Bug fixes must be submitted against the 1.12 branch
 - Features and deprecations must be submitted against the 1.13 branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

jakubtobiasz commented on src/Sylius/Bundle/OrderBundle/DependencyInjection/Compiler/TagCartContextsPass.php:
```diff
@@ -0,0 +1,39 @@
+            }
+        }
+    }
+}
```
New line is missing here.

jakubtobiasz commented on src/Sylius/Bundle/OrderBundle/DependencyInjection/Compiler/TagCartContextsPass.php:
```diff
@@ -0,0 +1,39 @@
+final class TagCartContextsPass implements CompilerPassInterface
+{
+
+    public function process(ContainerBuilder $container) : void
+    {
+        $this->tagWithKernelReset($container);
+    }
+
+    private function tagWithKernelReset(ContainerBuilder $container) : void
+    {
+        $taggedServices = $container->findTaggedServiceIds(RegisterCartContextsPass::CART_CONTEXT_SERVICE_TAG);
+
+        foreach ($taggedServices as $id => $tags) {
+            $definition = $container->getDefinition($id);
+
+            if(is_subclass_of($definition->getClass(), ResettingCartContextInterface::class)) {
+                $definition->addTag('kernel.reset', ['method' => 'reset']);
+            }
+        }
+    }
```
```suggestion
    public function process(ContainerBuilder $container) : void
    {
        $taggedServices = $container->findTaggedServiceIds(RegisterCartContextsPass::CART_CONTEXT_SERVICE_TAG);

        foreach ($taggedServices as $id => $tags) {
            $definition = $container->getDefinition($id);

            if(is_subclass_of($definition->getClass(), ResettingCartContextInterface::class)) {
                $definition->addTag('kernel.reset', ['method' => 'reset']);
            }
        }
    }
```

There's no need for an extra private method.

palttamas replied:
Then should the class name be `TagCartContextsWithKernelResetPass` ?

jakubtobiasz replied:
That's a good question :D. I looked at our existing Compiler Passes, and there's no schema we could apply here. From what I know, tomorrow more people will review this PR, so let's way for their opinion.

jakubtobiasz commented on src/Sylius/Bundle/OrderBundle/Tests/DependencyInjection/Compiler/TagCartContextsPassTest.php:
```diff
@@ -0,0 +1,94 @@
+    {
+        $container->addCompilerPass(new TagCartContextsPass());
+    }
+}
```
Missing new line

jakubtobiasz commented on src/Sylius/Bundle/OrderBundle/Tests/DependencyInjection/Compiler/TagCartContextsPassTest.php:
```diff
@@ -0,0 +1,94 @@
+final class TagCartContextsPassTest extends AbstractCompilerPassTestCase
+{
+
+    /**
+     * @test
+     */
```
```suggestion
    /** @test */
```
Small change to keep all tests consistent :). Applies to all test methods.

jakubtobiasz commented on src/Sylius/Component/Core/Cart/Context/ShopBasedCartContextInterface.php:
```diff
@@ -0,0 +1,21 @@
+use Sylius\Component\Order\Context\CartContextInterface;
+use Sylius\Component\Order\Context\ResettingCartContextInterface;
+
+interface ShopBasedCartContextInterface extends CartContextInterface, ResettingCartContextInterface
+{
+
+}
```
Missing new line

jakubtobiasz replied:
We're going to add the `reset` method on the `CartContextInterface` in the `2.0`, so I think we should remove this interface, and just declare `ResettingCartContextInterface` explicitly on `ShopBasedCartContext`.

jakubtobiasz commented on src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php:
```diff
@@ -26,7 +26,7 @@
 use Webmozart\Assert\Assert;
 
-final class ShopBasedCartContext implements CartContextInterface
+final class ShopBasedCartContext implements ShopBasedCartContextInterface
```
```suggestion
final class ShopBasedCartContext implements CartContextInterface, ResettingCartContextInterface
```

Explanation below 👇🏼.

jakubtobiasz commented on src/Sylius/Component/Order/Context/ResettingCartContextInterface.php:
```diff
@@ -0,0 +1,18 @@
+interface ResettingCartContextInterface
+{
+    public function reset(): void;
+}
```
Missig new line

jakubtobiasz commented on src/Sylius/Component/Order/Context/ResettingCartContextInterface.php:
```diff
@@ -0,0 +1,18 @@
+
+namespace Sylius\Component\Order\Context;
+
+interface ResettingCartContextInterface
```
As I wrote in other comment – in Sylius `2.0` we're going to add the `reset` method directly on the `CartContextInterface`, so let's add a deprecation note here. Just in case :P.

```suggestion
/** @deprecated since 1.13 and will be removed in Sylius 2.0. */
interface ResettingCartContextInterface
```

palttamas replied:
If you keep the `ResettingCartContextInterface` in 2.0 and make `CartContextInterface` extend it, then there is no need for the deprecation.

jakubtobiasz replied:
I know we can keep `ResettingCartContextInterface`, but we had many discussions how to handle it. AFAIR our ideal solution was to add this method to the `CartContextInterface`, but we couldn't due to BC policy.

jakubtobiasz reviewed (changes requested):
Hi 👋🏼!
Thanks for your contribution! I left you some comments (mostly cosmetic ones).

palttamas committed:
Abstracts out ShopBasedCartContext::reset to an interface

`ShopBasedCartContext` now implements the new `ResettingCartContextInterface`.

The service `sylius.context.cart.new_shop_based` is no longer explicitly tagged with the `kernel.reset` tag.

Cart contexts that implement the `ResettingCartContextInterface` will be automatically tagged with the `kernel.reset` tag.

palttamas commented:
@jakubtobiasz I think I made all the requested changes

NoResponseMate commented on psalm.xml:
```diff
@@ -59,6 +59,7 @@
                 <referencedClass name="Symfony\Component\Serializer\Normalizer\CacheableSupportsMethodInterface" /> <!-- deprecated in Symfony 6.3 -->
                 <referencedClass name="Symfony\Component\Serializer\Normalizer\ContextAwareDenormalizerInterface" /> <!-- deprecated in Symfony 6.1 -->
                 <referencedClass name="Symfony\Component\Serializer\Normalizer\ContextAwareNormalizerInterface" /> <!-- deprecated in Symfony 6.1 -->
+                <referencedClass name="Sylius\Component\Order\Context\ResettingCartContextInterface" /> <!-- deprecated in Sylius 2.0 -->
```
```suggestion
                <referencedClass name="Sylius\Component\Order\Context\ResettingCartContextInterface" /> <!-- deprecated in Sylius 1.13 -->
```
Although we could just drop the comment like in other internal deprecations.

palttamas replied:
The build doesn't pass, thus i had to add it

NoResponseMate replied:
I meant the comment alone, not the whole line

NoResponseMate commented on src/Sylius/Bundle/OrderBundle/DependencyInjection/Compiler/TagCartContextsPass.php:
```diff
@@ -0,0 +1,34 @@
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+final class TagCartContextsPass implements CompilerPassInterface
```
```suggestion
final class ResettingCartContextsPass implements CompilerPassInterface
```
?
At a glance, I'd assume that `TagCartContextsPass` is responsible for tagging services w/ the `sylius.context.cart` tag instead.

NoResponseMate commented on src/Sylius/Component/Order/Context/ResettingCartContextInterface.php:
```diff
@@ -0,0 +1,19 @@
+namespace Sylius\Component\Order\Context;
+
+/** @deprecated since 1.13 and will be removed in Sylius 2.0. */
+interface ResettingCartContextInterface
```
Couldn't we just use the `Symfony\Contracts\Service\ResetInterface` instead of creating a new, immediately deprecated one?

palttamas replied:
I would go as far to extend the ResetInterface, but i honestly don't see the benefit of using `ResetInterface` over a bespoke one.

```

class X {
    public function __construct(
        \Symfony\Contracts\Service\ResetInterface $resetableCartButUpToMeWhatToNameThis,
    )
    {

    }
}
```
vs

```
class X {
    public function __construct(
        Sylius\Component\Order\Context\ResettingCartContextInterface $canGiveThisParamAnyName,
    )
    {

    }
}

```

IMO the second usage expresses a lot more, using a generic interface makes the intent lost. I can't imagine a scenario where you would be applying the same calls to classes that implement `ResetInterface`. I'm curious what was the rationale behind introducing that at Symfony.

Edit: i guess it makes sense to extend the ResetInterface, in case something wants to grab all services that implement that and call reset, could be related to the `kernel.reset` tag?

palttamas replied:
Not arguing though, happy to make the change. Leave a thumbs down and I'll proceed with your suggestion

NoResponseMate replied:
Putting aside the fact that we're introducing a new interface that is immediately deprecated and presumably won't exist in the next major.
My main issue with creating this new interface is that the intent is the same: this service is stateful and should be reset. That's precisely what the `ResetInterface` is for, there's even [autoconfiguration](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L628) set on the framework extension that auto tags any service with this interface with `kernel.reset`.
IMO the Symfony logic in this case fits to a tee.

jakubtobiasz committed:
Remove custom ResettingCartContextInterface

jakubtobiasz committed:
Rename the pass and preventing tagging with the kernel.reset tag twice

jakubtobiasz commented:
I let myself implement @NoResponseMate suggestions as we'd like to merge this PR this week :).

jakubtobiasz committed:
Remove leftovers after removing ResettingCartContextInterface

jakubtobiasz committed:
Declare symfony/service-contracts dependency where needed

TheMilek commented:
Thank you, Tamas! :tada:

palttamas commented:
Sorry guys, haven't had time to look at this, thanks for merging :heart:

jakubtobiasz commented:
Np, we thank you for opening this PR 🫶🏼
